File: scada-front/src/features/DatabaseManager/Tags/types/Tag.ts
export abstract class Tag {
    tagName: string;
    tagType: string;
    description: string;
    ioAddress: number;

    constructor(tagName: string, tagType: string, description: string, ioAddress: number) {
        this.tagName = tagName;
        this.tagType = tagType;
        this.description = description;
        this.ioAddress = ioAddress;
    }
}

export class DigitalInputTag extends Tag {
    scanTime: number;
    scan: boolean;
    driver: string;

    constructor(tagName: string, tagType: string, description: string, ioAddress: number, scanTime: number, scan: boolean, driver: string) {
        super(tagName, tagType, description, ioAddress);
        this.scanTime = scanTime;
        this.scan = scan;
        this.driver = driver;
    }
}

export class DigitalOutputTag extends Tag {
    initialValue: number;

    constructor(tagName: string, tagType: string, description: string, ioAddress: number, initialValue: number) {
        super(tagName, tagType, description, ioAddress);
        this.initialValue = initialValue;
    }
}

export class AnalogInputTag extends Tag {
    scanTime: number;
    scan: boolean;
    lowLimit: number;
    highLimit: number;
    driver: string;
    units: string;

    constructor(
        tagName: string,
        tagType: string,
        description: string,
        ioAddress: number,
        scanTime: number,
        scan: boolean,
        lowLimit: number,
        highLimit: number,
        units: string,
        driver: string
    ) {
        super(tagName, tagType, description, ioAddress);
        this.scanTime = scanTime;
        this.scan = scan;
        this.lowLimit = lowLimit;
        this.highLimit = highLimit;
        this.units = units;
        this.driver = driver;
    }
}

export class AnalogOutputTag extends Tag {
    initialValue: number;
    lowLimit: number;
    highLimit: number;
    units: string;

    constructor(
        tagName: string,
        tagType: string,
        description: string,
        ioAddress: number,
        initialValue: number,
        lowLimit: number,
        highLimit: number,
        units: string
    ) {
        super(tagName, tagType, description, ioAddress);
        this.initialValue = initialValue;
        this.lowLimit = lowLimit;
        this.highLimit = highLimit;
        this.units = units;
    }
}

File: scada-back/Infrastructure/Feature/Tag/Model/Abstraction/Tag.cs
using System.Text.Json.Serialization;
using MongoDB.Bson;
using MongoDB.Bson.Serialization.Attributes;

namespace scada_back.Infrastructure.Feature.Tag.Model.Abstraction;

[BsonKnownTypes(typeof(AnalogInputTag), typeof(AnalogOutputTag), typeof(DigitalInputTag), typeof(DigitalOutputTag))]
public abstract class Tag
{
    [BsonId]
    [BsonRepresentation(BsonType.ObjectId)]
    public string Id { get; set; } = string.Empty;
    [BsonElement("tag_name")]
    public string TagName { get; set; } = string.Empty;
    [BsonElement("description")]
    public string Description { get; set; } = string.Empty;
    [BsonElement("io_address")]
    public int IOAddress { get; set; }

    public abstract TagDto ToDto();
}

public abstract class TagDto
{
    public string TagName { get; set; } = string.Empty;
    public string TagType { get; set; } = string.Empty;
    public string Description { get; set; } = string.Empty;
    public int IOAddress { get; set; }

    public abstract Tag ToEntity();

    public TagDto()
    {
        
    }
    
    [JsonConstructor]
    protected TagDto(string tagName, string tagType, string description, int ioAddress)
    {
        TagName = tagName;
        TagType = tagType;
        Description = description;
        IOAddress = ioAddress;
    }
}